# Kruskal-Wallis H Testi
Kruskal-Wallis, One-Way ANOVA'nın (Konu 69) normallik sağlanmadığındaki non-parametrik karşılığıdır. Mann-Whitney U Testi'nin (Konu, iki grup) **3 veya daha fazla gruba genişletilmiş** hali olarak da düşünülebilir.

## Ne Zaman Kullanılır?
3+ bağımsız grubu karşılaştırmak istediğimizde, ama en az bir grup normal dağılmadığında.

## Mantığı
Mann-Whitney gibi, Kruskal-Wallis de **tüm gözlemleri birleştirip sıralar**, sonra her grubun **ortalama sıralamasını (rank)** karşılaştırır. Eğer gruplardan biri sistematik olarak daha yüksek/düşük sıralarda kümeleniyorsa, bu farkı yakalar.

## Hipotezler
- **H0:** Tüm grupların dağılımları aynıdır (medyanları eşittir)
- **H1:** En az bir grubun dağılımı diğerlerinden farklıdır

(One-Way ANOVA'daki gibi, Kruskal-Wallis de sadece "en az biri farklı" der, hangisinin farklı olduğunu söylemez — anlamlı çıkarsa,Post-Hoc'un non-parametrik karşılığı olan **Dunn's Test** ile takip edilir, ismini bilmemiz yeterli.)

## Python'da Kullanımı
```python
from scipy.stats import kruskal

istatistik, p_degeri = kruskal(grup1, grup2, grup3)
```

## Test İstatistiği
- H istatistiği, Ki-Kare dağılımına göre yorumlanır (F dağılımına göre değil, ANOVA'dan farklı olarak).
- serbestlik derecesi = grup sayısı - 1

In [4]:
import numpy as np
from scipy import stats
np.random.seed(42)

telefon = np.random.exponential(scale=20, size=30)
canli_sohbet = np.random.exponential(scale=12, size=30)
email = np.random.exponential(scale=35, size=30)

# H0: Veriler normal dağılıma uyuyor.
# H1: Veriler normal dağılıma uymuyor.

telefon_normallik = stats.shapiro(telefon).pvalue
canli_sohbet_normallik = stats.shapiro(canli_sohbet).pvalue
email_normallik = stats.shapiro(email).pvalue
alpha = 0.05
if telefon_normallik < alpha:
    print('Telefon kanalının çözüm süresi normal dağılmıyor')
else:
    print('Telefon kanalının çözüm süresi normal dağılıyor')
if canli_sohbet_normallik < alpha:
    print('Canlı sohbet kanalının çözüm süresi normal dağılmıyor.')
else:
    print('Canlı sohbet kanalının çözüm süresi normal dağılıyor.')
if email_normallik < alpha:
    print('Email kanalının çözüm süresi normal dağılmıyor.')
else:
    print('Email kanalının çözüm süresi normal dağılıyor.')

Telefon kanalının çözüm süresi normal dağılmıyor
Canlı sohbet kanalının çözüm süresi normal dağılmıyor.
Email kanalının çözüm süresi normal dağılmıyor.


Her 3 grup da normal dağılmadığı için ANOVA testinin parametrik olmayan versiyonu olan Kruskal-Wallis H testini uygulayacağız.

In [5]:
# Kruskal-Wallis H testi için hipotez yapısı
# H0: Tüm grupların dağılımı aynıdır.
# H1: En az bir grubun dağılımı diğer gruplarından farklıdır.
istatistik, p_degeri = stats.kruskal(telefon, canli_sohbet, email)
print(f'H istatistiği: {istatistik}')
print(f'Kruskal-Wallis H testine ait P değeri: {p_degeri}')
if p_degeri < alpha:
    print('En az bir grubun dağılımı diğer gruplardan farklıdır.')
else:
    print('Tüm grupların dağılımı aynıdır.')
telefon_ortalama = np.mean(telefon)
canli_sohbet_ortalama = np.mean(canli_sohbet)
email_ortalama = np.mean(email)
telefon_medyan = np.median(telefon)
canli_sohbet_medyan = np.median(canli_sohbet)
email_medyan = np.median(email)
print('Telefon kanalına ait şikayetlerin çözüm sürelerinin ortalaması: {}'.format(telefon_ortalama))
print('Canlı Sohbet kanalına ait şikayetlerin çözüm sürelerinin ortalaması: {}'.format(canli_sohbet_ortalama))
print('Email kanalına ait şikayetlerin çözüm sürelerinin ortalaması: {}'.format((email_ortalama)))
print(f'Telefon kanalına ait şikayetlerin çözüm sürelerinin medyanı: {telefon_medyan}')
print(f'Canlı Sohbet kanalına ait şikayetlerin çözüm sürelerinin medyanı: {canli_sohbet_medyan}')
print(f'Email kanalına ait şikayetlerin çözüm sürelerinin medyanı: {email_medyan}')

H istatistiği: 9.552039072039065
Kruskal-Wallis H testine ait P değeri: 0.008429485540797366
En az bir grubun dağılımı diğer gruplardan farklıdır.
Telefon kanalına ait şikayetlerin çözüm sürelerinin ortalaması: 16.11600612132058
Canlı Sohbet kanalına ait şikayetlerin çözüm sürelerinin ortalaması: 12.950642855502837
Email kanalına ait şikayetlerin çözüm sürelerinin ortalaması: 32.318178266680924
Telefon kanalına ait şikayetlerin çözüm sürelerinin medyanı: 10.348051567571957
Canlı Sohbet kanalına ait şikayetlerin çözüm sürelerinin medyanı: 8.505948744260273
Email kanalına ait şikayetlerin çözüm sürelerinin medyanı: 19.795937440484245


Kruskal-Wallis H testi sonuçlarına göre, yukarıdaki 3 kanalın çözüm sürelerinden (telefon, canlı sohbet, email) en az bir tanesinin dağılımı diğerlerinden anlamlı biçimde farklıdır ancak hangi grupların birbirinden farklılaştığını görmemiz için Post-Hoc (Çoklu karşılaştırma) testi uygulamamız gerekir ama biz bu testi şu an için uygulamayacağız. İlgili kanallara ait tanımlayıcı istatistikleri incelediğimizde, Email kanalına ait şikayetlerin çözüm sürelerinin ortalaması (32.32) > Telefon kanalına ait şikayetlerin çözüm sürelerinin ortalaması (16.12) > Canlı sohbet kanalına ait şikayetlerin çözüm sürelerinin ortalaması (12.95) şeklinde iken Medyan değerlerini incelediğimizde ise, Email kanalına ait şikayetlerin çözüm sürelerinin medyan değeri (19.8) > Telefon kanalına ait şikayetlerin çözüm sürelerinin medyan değeri (10.35) > Canlı Sohbet kanalına ait şikayetlerin çözüm sürelerinin medyan değeri (8.51) şeklinde sıralanmaktadır. 